src/github.com/couchbase/sync_gateway/base/sgw_pindex.go

```diff
@@ -0,0 +1,316 @@
+	return c.eventFeed
+}
+
+func (c *CBGTDCPFeed) Close() error { // TODO
+	log.Fatalf("CBGTDCPFeed.Close() called but not implemented")
+	return nil
+}
+
+type SyncGatewayPIndex struct {
+	mutex        sync.Mutex             // mutex used to protect meta and seqs
+	seqs         map[uint16]uint64      // To track max seq #'s we received per partition (vbucketId).
+	meta         map[uint16][]byte      // To track metadata blob's per partition (vbucketId).
+	feedEvents   chan sgbucket.TapEvent // The channel to forward TapEvents
+	bucket       CouchbaseBucket        // the couchbase bucket
+	tapArguments sgbucket.TapArguments  // tap args
+}
+
+func NewSyncGatewayPIndex(feedEvents chan sgbucket.TapEvent, bucket CouchbaseBucket, args sgbucket.TapArguments) *SyncGatewayPIndex {
+	pindex := &SyncGatewayPIndex{
+		feedEvents:   feedEvents,
+		bucket:       bucket,
+		tapArguments: args,
+	}
+
+	if err := pindex.SeedSeqnos(); err != nil {
+		log.Fatalf("Error calling SeedSeqnos for pindex: %v", err)
+	}
+
+	return pindex
+}
+
+func (s *SyncGatewayPIndex) SeedSeqnos() error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	maxVbno, err := s.bucket.GetMaxVbno()
+	if err != nil {
+		return err
+	}
+
+	startSeqnos := make(map[uint16]uint64, maxVbno)
+	vbuuids := make(map[uint16]uint64, maxVbno)
+
+	// GetStatsVbSeqno retrieves high sequence number for each vbucket, to enable starting
+	// DCP stream from that position.  Also being used as a check on whether the server supports
+	// DCP.
+	statsUuids, highSeqnos, err := s.bucket.GetStatsVbSeqno(maxVbno)
+	if err != nil {
+		return errors.New("Error retrieving stats-vbseqno - DCP not supported")
+	}
+
+	if s.tapArguments.Backfill == sgbucket.TapNoBackfill {
+		// For non-backfill, use vbucket uuids, high sequence numbers
+		LogTo("Feed+", "Seeding seqnos: %v", highSeqnos)
+		vbuuids = statsUuids
+		startSeqnos = highSeqnos
+	}
+
+	// Set the high seqnos as-is
+	s.seqs = startSeqnos
+
+	// For metadata, we need to do more work to build metadata based on uuid and map values.  This
+	// isn't strictly to the design of cbdatasource.Receiver, which intends metadata to be opaque, but
+	// is required in order to have the BucketDataSource start the UPRStream as needed.
+	// The implementation has been reviewed with the cbdatasource owners and they agree this is a
+	// reasonable approach, as the structure of VBucketMetaData is expected to rarely change.
+	for vbucketId, vbuuid := range vbuuids {
+		failOver := make([][]uint64, 1)
+		failOverEntry := []uint64{vbuuid, 0}
+		failOver[0] = failOverEntry
+		metadata := &cbdatasource.VBucketMetaData{
+			SeqStart:    s.seqs[vbucketId],
+			SeqEnd:      uint64(0xFFFFFFFFFFFFFFFF),
+			SnapStart:   s.seqs[vbucketId],
+			SnapEnd:     s.seqs[vbucketId],
+			FailOverLog: failOver,
+		}
+		buf, err := json.Marshal(metadata)
+		if err == nil {
+			if s.meta == nil {
+				s.meta = make(map[uint16][]byte)
+			}
+			s.meta[vbucketId] = buf
+		}
+	}
```

Comment:
Are we still seeding with the high seqnos from the vbstats call?  Should be using the index bucket stable sequence - maybe that work is still pending?